File: autoviz/examples/cpp/11_tutorial_state_transitions.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <automsgs/msgs/std_msgs/string.pb.h>

#include "common/tutorial_utils.hpp"

namespace {

using Step = std::pair<int, const char*>;

const Step kNav[] = {
    {4, "IDLE"},     {3, "PLANNING"}, {6, "NAVIGATING"},
    {2, "RECOVERY"}, {5, "NAVIGATING"}, {3, "IDLE"},
};
const Step kCtrl[] = {
    {5, "IDLE"}, {8, "TRACKING"}, {3, "HOLD"}, {6, "TRACKING"}, {2, "IDLE"},
};
const Step kMode[] = {
    {10, "MANUAL"}, {12, "AUTO"}, {4, "ESTOP"}, {8, "AUTO"},
};

template <size_t N>
std::string Advance(const Step (&seq)[N], int* i, int* hold) {
  const std::string state = seq[*i].second;
  --(*hold);
  if (*hold <= 0) {
    *i = (*i + 1) % static_cast<int>(N);
    *hold = seq[*i].first;
  }
  return state;
}

}  // namespace

using autoviz::examples::MakeWriter;
using autoviz::examples::ParseRate;

int main(int argc, char** argv) {
  const double rate_hz = ParseRate(argc, argv, 2.0);

  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/state");
  auto w_nav =
      MakeWriter<automsgs::msgs::std_msgs::String>(node, "/fake/state/nav");
  auto w_ctrl =
      MakeWriter<automsgs::msgs::std_msgs::String>(node, "/fake/state/ctrl");
  auto w_mode =
      MakeWriter<automsgs::msgs::std_msgs::String>(node, "/fake/state/mode");

  autolink::Rate rate(rate_hz);
  std::cout << "state @ " << rate_hz
            << " Hz → /fake/state/{nav,ctrl,mode}\n";

  int ni = 0, ci = 0, mi = 0;
  int nh = kNav[0].first, ch = kCtrl[0].first, mh = kMode[0].first;

  while (autolink::OK()) {
    const std::string nav = Advance(kNav, &ni, &nh);
    const std::string ctrl = Advance(kCtrl, &ci, &ch);
    const std::string mode = Advance(kMode, &mi, &mh);

    auto m_nav = std::make_shared<automsgs::msgs::std_msgs::String>();
    m_nav->set_data(nav);
    w_nav->Write(m_nav);

    auto m_ctrl = std::make_shared<automsgs::msgs::std_msgs::String>();
    m_ctrl->set_data(ctrl);
    w_ctrl->Write(m_ctrl);

    auto m_mode = std::make_shared<automsgs::msgs::std_msgs::String>();
    m_mode->set_data(mode);
    w_mode->Write(m_mode);

    rate.Sleep();
  }
  return 0;
}

File: autoviz/examples/python/11_tutorial_state_transitions.py
#!/usr/bin/env python3
"""Publish discrete states for Autoviz State Transitions panel.

  /fake/state/nav   std_msgs/String  → field data
  /fake/state/ctrl  std_msgs/String  → field data
  /fake/state/mode  std_msgs/String  → field data

Add series: channel + Message path `data`.

  /usr/bin/python3 examples/python/11_tutorial_state_transitions.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import argparse

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink
from automsgs.msgs.std_msgs import string_pb2

# (hold_ticks, state) — each step holds for hold_ticks at --rate
NAV = (
    (4, 'IDLE'),
    (3, 'PLANNING'),
    (6, 'NAVIGATING'),
    (2, 'RECOVERY'),
    (5, 'NAVIGATING'),
    (3, 'IDLE'),
)
CTRL = (
    (5, 'IDLE'),
    (8, 'TRACKING'),
    (3, 'HOLD'),
    (6, 'TRACKING'),
    (2, 'IDLE'),
)
MODE = (
    (10, 'MANUAL'),
    (12, 'AUTO'),
    (4, 'ESTOP'),
    (8, 'AUTO'),
)


def advance(seq, i, hold):
    """Return (state, next_i, next_hold). `hold` includes the current tick."""
    state = seq[i][1]
    hold -= 1
    if hold <= 0:
        i = (i + 1) % len(seq)
        hold = seq[i][0]
    return state, i, hold


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--rate', type=float, default=2.0)
    args = ap.parse_args()

    autolink.init('state')
    n = autolink.Node('/autoviz/state')
    w_nav = n.create_writer('/fake/state/nav', string_pb2.String, qos_depth=10)
    w_ctrl = n.create_writer('/fake/state/ctrl', string_pb2.String, qos_depth=10)
    w_mode = n.create_writer('/fake/state/mode', string_pb2.String, qos_depth=10)
    rate = autolink.Rate(args.rate)
    print(f'state @ {args.rate} Hz → /fake/state/{{nav,ctrl,mode}}')

    ni = ci = mi = 0
    nh, ch, mh = NAV[0][0], CTRL[0][0], MODE[0][0]
    try:
        while not autolink.is_shutdown():
            nav, ni, nh = advance(NAV, ni, nh)
            ctrl, ci, ch = advance(CTRL, ci, ch)
            mode, mi, mh = advance(MODE, mi, mh)
            w_nav.write(string_pb2.String(data=nav))
            w_ctrl.write(string_pb2.String(data=ctrl))
            w_mode.write(string_pb2.String(data=mode))
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()
